api/STRUCTURE.md

```diff
@@ -0,0 +1,83 @@
+
+TODO: wording on the above is Bad. pls help
+
+Much of the structure of DB models can be autogenerated in IntelliJ by defining the class members and then doing Generate -> Getter and Setter and Generate -> `equals()` and `hashCode()`. This will not automatically add the required annotations - those must be added manually.
+
+## Services and ServiceImpls
+
+Services are interfaces that declare methods of business logic. ServiceImpls are classes implementing their correspondingly named Service and the business logic methods declared therein.
+
+TODO: Why do we do this this way?
+
+Services / ServiceImpls are generally named after a feature or functional area, like 'UserService', 'CohortService', or 'MonitoringService'.
+
+ServiceImpls can call into other services, but we try to keep the number of dependencies down. We try to avoid having ServiceImpls call into a large number of DAOs, especially DAOs not directly related to the functional area for which the ServiceImpl is responsible. For instance, the `UserServiceImpl` calls into the `UserDao`, the `AdminActionHistoryDao`
+
+ServiceImpls should handle only application logic. They should not need explicit knowledge of the database, the REST API, or the UI.
+
+## Controllers
+
+Controllers take Request objects from the API, call into the Service layer, and return Response objects back to the API. Controllers implement autogenerated [APIControllers](#apicontrollers).
+
+Controller methods should generally be small and should not contain any actual logic. The most common conditional in the Controller layer should be checking a feature flag. Most of the work done by Controllers should be munging data provided by the API or a Service.
+
+### Mappers
+
+There are four types of models in the system - [API models](#swagger), [DB models](#db-models), and [API Request and Response](#requests-and-responses) models. Mappers are utility classes that own methods that convert between these types of models.
+
+## API
+
+The actual REST API is defined in a series of YAML files according to the [Swagger OpenAPI spec](https://swagger.io/specification/). Swagger takes these YAML files, merges them, and generates all the classes defined by the YAMLs. The command to do so is `./project.rb compile-generated-java`.
+
+One non-Swagger-standard thing about these YAML files is that the `tag` param is used to indicate which Controller / APIController an endpoint should be associated with.
+
+Each REST endpoint is defined under `PATHS` in the YAML file.
+
+### API Models
+
+API models are defined under `DEFINITIONS` in the YAML file. Swagger autogenerates classes from these. Both API models and DB models are currently used in the Controller and Service layer. There is not currently consensus on whether the standard should switch to only using one or the other.
+
+TODO: is the latter sentence true?
```

Comment:
The major advantage of separating these is allowing a more layered architecture and stricter separation of business logic from database details. Granted, Spring is already giving us some abstraction.

Reply:
I understand this. I'm just asking whether we as a team actually agree on what types of models we should be using where. We certainly haven't prioritized conforming to any particular standard.

Reply:
There's a lot of value in what you're doing, i.e. to describe in some systematic way the general understanding of how things work today. It makes it possible to lay out a workable plan for improving things.